index.js

```diff
@@ -81,11 +81,20 @@ function restoreEnv(origEnv) {
 function callbackInModifiedEnv(callback, setInEnv, removeFromEnv) {
   var origEnv = {};
   var result;
+
   setVars(origEnv, setInEnv);
   delVars(origEnv, removeFromEnv);
+
   result = callback();
-  restoreEnv(origEnv);
-  return result;
+
+  if (result && typeof result.then === 'function') {
```

Comment:
I think this should use `Promise.resolve()` when it's available, instead of just trusting the `then` function.